importer/import.go

```diff
@@ -0,0 +1,282 @@
+	}
+
+	metricSamples, minValidTimestamp, maxValidTimestamp, err := parseMetrics(bytes, contentType)
+	if err != nil {
+		return err
+	}
+
+	if !skipTimestampCheck {
+		err = verifyIntegration(dbPath, minValidTimestamp, maxValidTimestamp)
+		if err != nil {
+			return err
+		}
+	}
+
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	//snapshotPath, err := pushToDisk(metricSamples, tmpDbDir, minValidTimestamp, logger)
+	snapshotPaths, err := pushToDisk(metricSamples, tmpDbDir, logger)
+	if err != nil {
+		return err
+	}
+	level.Info(logger).Log("msg", "blocks created", "blockPaths", snapshotPaths)
+
+	err = copyToDatabase(tmpDbDir, dbPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// parseMetrics parses metrics formatted in the Prometheus exposition format.
+// Returns the metric samples, min timestamp, max timestamp, and error.
+func parseMetrics(b []byte, contentType string) ([][]*metricSample, timestamp, timestamp, error) {
+	var minValidTimestamp timestamp
+	minValidTimestamp = math.MaxInt64
+	var maxValidTimestamp timestamp
+	maxValidTimestamp = math.MinInt64
+
+	var buckets [][]*metricSample
+	var currentBucket []*metricSample
+	var startTime time.Time
+	var err error
+	parser := textparse.New(b, contentType)
+	for {
+		var ent textparse.Entry
+		if ent, err = parser.Next(); err != nil {
+			// Error strings are just differently enough across packages,
+			// hence this catch-all that just looks for "EOF" in the error
+			// string, and if it finds one, it means that the parsing is complete.
+			if strings.Contains(strings.ToLower(err.Error()), "eof") {
+				err = nil
+				break
+			}
+			// In case the error that we see is not related to the EOF.
+			return nil, minValidTimestamp, maxValidTimestamp, err
+		}
+		switch ent {
+		case textparse.EntryType:
+			continue
+		case textparse.EntryHelp:
+			continue
+		case textparse.EntryUnit:
+			continue
+		case textparse.EntryComment:
+			continue
+		default:
+		}
+		_, currentTimestampNs, val := parser.Series()
+
+		// Parses converts all timestamps to nanoseconds.
+		// TSDB looks for timestamps in milliseconds.
+		var currentTimestampMs timestamp
+		if currentTimestampNs == nil {
+			currentTimestampMs = 0
+		} else {
+			currentTimestampMs = *currentTimestampNs / 1000
+		}
+
+		minValidTimestamp = minInt(minValidTimestamp, currentTimestampMs)
+		maxValidTimestamp = maxInt(maxValidTimestamp, currentTimestampMs)
+
+		var lset prom_labels.Labels
+		_ = parser.Metric(&lset)
+
+		tsdbLabels := tsdb_labels.FromMap(lset.Map())
+
+		currentTime := time.Unix(currentTimestampMs/1000, 0)
+		currentSample := &metricSample{TimestampMs: currentTimestampMs, Value: val, Labels: tsdbLabels}
+
+		if startTime.IsZero() {
+			startTime = currentTime
+			currentBucket = append(currentBucket, currentSample)
+			continue
+		}
+
+		timeDelta := currentTime.Sub(startTime)
+		if timeDelta.Seconds()*1000 >= BlockDuration {
```

Comment:
What if goes back a bucket?

Reply:
You mean if the samples are out of order?

Based on the exporter I've written above, the input to the parser is expected to be chronologically sorted in ascending order. We could have guard function that first ingests all the samples, and then sorts them according to the timestamp.

Reply:
I mean what if you have one metric that spans buckets (in order), and then a second that spans buckets (also in order).

Reply:
I'm not quite sure what you mean by that. Can you give me an example?

Reply:
a{foo="bar"} 0 1
a{foo="bar"} 0 100000000000
b{foo="bar"} 0 1
b{foo="bar"} 0 100000000000

Reply:
So in this case, it throws an out of bounds error, if we go back a bucket.